File: parser/src/structure.rs
use serde::{Deserialize, Serialize};

use super::name::Name;
use super::types::Type;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SysDCSystem {
    pub units: Vec<SysDCUnit>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SysDCUnit {
    pub name: Name,
    pub data: Vec<SysDCData>,
    pub modules: Vec<SysDCModule>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SysDCData {
    pub name: Name,
    pub members: Vec<(Name, Type)>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SysDCModule {
    pub name: Name,
    pub functions: Vec<SysDCFunction>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SysDCFunction {
    pub name: Name,
    pub args: Vec<(Name, Type)>,
    pub returns: (Name, Type),
    pub annotations: Vec<SysDCAnnotation>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum SysDCAnnotation {
    Affect {
        func: (Name, Type),
        args: Vec<(Name, Type)>,
    },
    Modify {
        target: (Name, Type),
        uses: Vec<(Name, Type)>,
    },
    Spawn {
        result: (Name, Type),
        details: Vec<SysDCSpawnDetail>,
    },
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum SysDCSpawnDetail {
    Use(Name, Type),
    Return(Name, Type),
    LetTo {
        name: Name,
        func: (Name, Type),
        args: Vec<(Name, Type)>,
    },
}

pub mod unchecked {
    use anyhow;

    use super::Name;
    use super::Type;

    #[derive(Debug)]
    pub struct SysDCSystem {
        pub units: Vec<SysDCUnit>,
    }

    impl SysDCSystem {
        pub fn new(units: Vec<SysDCUnit>) -> SysDCSystem {
            SysDCSystem { units }
        }

        pub fn convert<F>(self, mut converter: F) -> anyhow::Result<super::SysDCSystem>
        where
            F: FnMut(SysDCUnit) -> anyhow::Result<super::SysDCUnit>,
        {
            let mut units = vec![];
            for unit in self.units {
                units.push(converter(unit)?);
            }
            Ok(super::SysDCSystem { units })
        }
    }

    #[derive(Debug)]
    pub struct SysDCUnit {
        pub name: Name,
        pub data: Vec<SysDCData>,
        pub modules: Vec<SysDCModule>,
        pub imports: Vec<Name>,
    }

    impl SysDCUnit {
        pub fn new(
            name: Name,
            data: Vec<SysDCData>,
            modules: Vec<SysDCModule>,
            imports: Vec<Name>,
        ) -> SysDCUnit {
            SysDCUnit {
                name,
                data,
                modules,
                imports,
            }
        }

        pub fn convert<F, G>(
            self,
            d_converter: F,
            m_converter: G,
        ) -> anyhow::Result<super::SysDCUnit>
        where
            F: Fn(SysDCData) -> anyhow::Result<super::SysDCData>,
            G: Fn(SysDCModule) -> anyhow::Result<super::SysDCModule>,
        {
            let (mut data, mut modules) = (vec![], vec![]);
            for _data in self.data {
                data.push(d_converter(_data)?);
            }
            for module in self.modules {
                modules.push(m_converter(module)?);
            }
            Ok(super::SysDCUnit {
                name: self.name,
                data,
                modules,
            })
        }
    }

    #[derive(Debug)]
    pub struct SysDCData {
        pub name: Name,
        pub members: Vec<(Name, Type)>,
    }

    impl SysDCData {
        pub fn new(name: Name, members: Vec<(Name, Type)>) -> SysDCData {
            SysDCData { name, members }
        }

        pub fn convert<F>(self, converter: F) -> anyhow::Result<super::SysDCData>
        where
            F: Fn((Name, Type)) -> anyhow::Result<(Name, Type)>,
        {
            let mut members = vec![];
            for member in self.members {
                members.push(converter(member)?);
            }
            Ok(super::SysDCData {
                name: self.name,
                members,
            })
        }
    }

    #[derive(Debug)]
    pub struct SysDCModule {
        pub name: Name,
        pub functions: Vec<SysDCFunction>,
    }

    impl SysDCModule {
        pub fn new(name: Name, functions: Vec<SysDCFunction>) -> SysDCModule {
            SysDCModule { name, functions }
        }

        pub fn convert<F>(self, converter: F) -> anyhow::Result<super::SysDCModule>
        where
            F: Fn(SysDCFunction) -> anyhow::Result<super::SysDCFunction>,
        {
            let mut functions = vec![];
            for func in self.functions {
                functions.push(converter(func)?);
            }
            Ok(super::SysDCModule {
                name: self.name,
                functions,
            })
        }
    }

    #[derive(Debug)]
    pub struct SysDCFunction {
        pub name: Name,
        pub args: Vec<(Name, Type)>,
        pub returns: (Name, Type),
        pub annotations: Vec<SysDCAnnotation>,
    }

    impl SysDCFunction {
        pub fn new(
            name: Name,
            args: Vec<(Name, Type)>,
            returns: (Name, Type),
            annotations: Vec<SysDCAnnotation>,
        ) -> SysDCFunction {
            SysDCFunction {
                name,
                args,
                returns,
                annotations,
            }
        }

        pub fn convert<F, G, H>(
            self,
            a_convert: F,
            r_convert: G,
            s_convert: H,
        ) -> anyhow::Result<super::SysDCFunction>
        where
            F: Fn((Name, Type)) -> anyhow::Result<(Name, Type)>,
            G: Fn((Name, Type)) -> anyhow::Result<(Name, Type)>,
            H: Fn(SysDCAnnotation) -> anyhow::Result<super::SysDCAnnotation>,
        {
            let returns = r_convert(self.returns)?;
            let mut args = vec![];
            let mut annotations = vec![];
            for arg in self.args {
                args.push(a_convert(arg)?);
            }
            for annotation in self.annotations {
                annotations.push(s_convert(annotation)?);
            }
            Ok(super::SysDCFunction {
                name: self.name,
                args,
                returns,
                annotations,
            })
        }
    }

    #[derive(Debug, Clone)]
    pub enum SysDCAnnotation {
        Return(Name),
        Affect {
            func: (Name, Type),
            args: Vec<(Name, Type)>,
        },
        Modify {
            target: (Name, Type),
            uses: Vec<(Name, Type)>,
        },
        Spawn {
            result: (Name, Type),
            details: Vec<SysDCSpawnDetail>,
        },
    }

    impl SysDCAnnotation {
        pub fn new_return(name: Name) -> SysDCAnnotation {
            SysDCAnnotation::Return(name)
        }

        pub fn new_affect(func: (Name, Type), args: Vec<(Name, Type)>) -> SysDCAnnotation {
            SysDCAnnotation::Affect { func, args }
        }

        pub fn new_modify(target: (Name, Type), uses: Vec<(Name, Type)>) -> SysDCAnnotation {
            SysDCAnnotation::Modify { target, uses }
        }

        pub fn new_spawn(result: (Name, Type), details: Vec<SysDCSpawnDetail>) -> SysDCAnnotation {
            SysDCAnnotation::Spawn { result, details }
        }

        pub fn convert<F, G, H>(
            self,
            a_converter: F,
            m_converter: G,
            s_converter: H,
        ) -> anyhow::Result<super::SysDCAnnotation>
        where
            F: Fn(
                (Name, Type),
                Vec<(Name, Type)>,
            ) -> anyhow::Result<((Name, Type), Vec<(Name, Type)>)>,
            G: Fn(
                (Name, Type),
                Vec<(Name, Type)>,
            ) -> anyhow::Result<((Name, Type), Vec<(Name, Type)>)>,
            H: Fn(
                (Name, Type),
                Vec<SysDCSpawnDetail>,
            ) -> anyhow::Result<((Name, Type), Vec<super::SysDCSpawnDetail>)>,
        {
            match self {
                SysDCAnnotation::Affect { func, args } => {
                    let (func, args) = a_converter(func, args)?;
                    Ok(super::SysDCAnnotation::Affect { func, args })
                }
                SysDCAnnotation::Modify { target, uses } => {
                    let (target, uses) = m_converter(target, uses)?;
                    Ok(super::SysDCAnnotation::Modify { target, uses })
                }
                SysDCAnnotation::Spawn { result, details } => {
                    let (result, details) = s_converter(result, details)?;
                    Ok(super::SysDCAnnotation::Spawn { result, details })
                }
                _ => panic!("Internal error"),
            }
        }
    }

    #[derive(Debug, Clone)]
    pub enum SysDCSpawnDetail {
        Use(Name, Type),
        Return(Name, Type),
        LetTo {
            name: Name,
            func: (Name, Type),
            args: Vec<(Name, Type)>,
        },
    }

    impl SysDCSpawnDetail {
        pub fn new_use(name: Name, types: Type) -> SysDCSpawnDetail {
            SysDCSpawnDetail::Use(name, types)
        }

        pub fn new_return(name: Name, types: Type) -> SysDCSpawnDetail {
            SysDCSpawnDetail::Return(name, types)
        }

        pub fn new_let_to(
            name: Name,
            func: (Name, Type),
            args: Vec<(Name, Type)>,
        ) -> SysDCSpawnDetail {
            SysDCSpawnDetail::LetTo { name, func, args }
        }

        pub fn convert<F, G>(
            self,
            u_converter: F,
            r_converter: F,
            l_converter: G,
        ) -> anyhow::Result<super::SysDCSpawnDetail>
        where
            F: Fn((Name, Type)) -> anyhow::Result<(Name, Type)>,
            G: Fn(
                Name,
                (Name, Type),
                Vec<(Name, Type)>,
            ) -> anyhow::Result<(Name, (Name, Type), Vec<(Name, Type)>)>,
        {
            match self {
                SysDCSpawnDetail::Use(name, types) => {
                    let (name, types) = u_converter((name, types))?;
                    Ok(super::SysDCSpawnDetail::Use(name, types))
                }
                SysDCSpawnDetail::Return(name, types) => {
                    let (name, types) = r_converter((name, types))?;
                    Ok(super::SysDCSpawnDetail::Return(name, types))
                }
                SysDCSpawnDetail::LetTo { name, func, args } => {
                    let (name, func, args) = l_converter(name, func, args)?;
                    Ok(super::SysDCSpawnDetail::LetTo { name, func, args })
                }
            }
        }
    }
}

File: tools/view/src/structure.ts
/* structure definitions */

export type Name = string;

export type Type = string;

export type SysDCSystem = {
    readonly units: SysDCUnit[]
}

export type SysDCUnit = {
    readonly name: Name,
    readonly data: SysDCData[],
    readonly modules: SysDCModule[]
}

export type SysDCData = {
    readonly name: Name,
    readonly members: [Name, Type][]
}

export type SysDCModule = {
    readonly name: Name,
    readonly functions: SysDCFunction[]
}

export type SysDCFunction = {
    readonly name: Name,
    readonly args: [Name, Type][],
    readonly return: [Name, Type],
    readonly annotations: SysDCAnnotation[]
}

export type SysDCAnnotation =
    SysDCAnnotationAffect |
    SysDCAnnotationModify |
    SysDCAnnotationSpawn

export type SysDCAnnotationAffect = {
    readonly func: [Name, Type],
    readonly args: [Name, Type][]
}

export type SysDCAnnotationModify = {
    readonly target: [Name, Type],
    readonly uses: [Name, Type][]
}

export type SysDCAnnotationSpawn = {
    readonly result: [Name, Type],
    readonly details: SysDCSpawnDetail[]
};

export type SysDCSpawnDetail =
    SysDCSpawnDetailUse |
    SysDCSpawnDetailReturn |
    SysDCSpawnDetailLetTo

export type SysDCSpawnDetailUse = readonly [Name, Type]

export type SysDCSpawnDetailReturn = readonly [Name, Type]

export type SysDCSpawnDetailLetTo = {
    readonly name: Name,
    readonly func: [Name, Type],
    readonly args: [Name, Type][]
}

/* converters */

export function convert(obj: object): SysDCSystem {
    return {
        units: obj["units"].map(__convert_unit)
    };
}

function __convert_unit(obj: object): SysDCUnit {
    return {
        name: __convert_name(obj["name"]),
        data: obj["data"].map(__convert_data),
        modules: obj["modules"].map(__convert_module)
    };
}

function __convert_data(obj: object): SysDCData {
    return {
        name: __convert_name(obj["name"]),
        members: obj["members"].map(__convert_name_type)
    };
}

function __convert_module(obj: object): SysDCModule {
    return {
        name: __convert_name(obj["name"]),
        functions: obj["functions"].map(__convert_function)
    };
}

function __convert_function(obj: object): SysDCFunction  {
    return {
        name: __convert_name(obj["name"]),
        args: obj["args"].map(__convert_name_type),
        return: __convert_name_type(obj["returns"]),
        annotations: obj["annotations"].map(__convert_annotation)
    };
}

function __convert_annotation(obj: object): SysDCAnnotation {
    if (obj["Affect"] != undefined) {
        return {
            func: __convert_name_type(obj["Affect"]["func"]),
            args: obj["Affect"]["args"].map(__convert_name_type)
        };
    }
    if (obj["Modify"] != undefined) {
        return {
            target: __convert_name_type(obj["Modify"]["target"]),
            uses: obj["Modify"]["uses"].map(__convert_name_type)
        };
    }
    if (obj["Spawn"] != undefined) {
        return {
            result: __convert_name_type(obj["Spawn"]["result"]),
            details: obj["Spawn"]["details"].map(__convert_spawn_detail)
        };
    }
}

function __convert_spawn_detail(obj: Object): SysDCSpawnDetail {
    if (obj["Use"] != undefined) {
        return __convert_name_type(obj["Use"]);
    }
    if (obj["Return"] != undefined) {
        return __convert_name_type(obj["Return"]);
    }
    if (obj["LetTo"] != undefined) {
        return {
            name: __convert_name(obj["LetTo"]["name"]),
            func: __convert_name_type(obj["LetTo"]["func"]),
            args: obj["LetTo"]["args"].map(__convert_name_type)
        };
    }
}

function __convert_name(obj: object): Name {
    return obj["namespace"] + "." + obj["name"];
}

function __convert_type(obj: object): Type {
    if (obj["refs"] == null) {
        return obj["kind"];
    }
    return __convert_name(obj["refs"]);
}

function __convert_name_type(obj: object): [Name, Type] {
    return [__convert_name(obj[0]), __convert_type(obj[1])];
}
